### PR TITLE
Added the ability to store important output from cli

### PR DESCRIPTION
**Pull request by JacobCallahan**
Added Emitter class and emit instance to helpers
Added a way to gain failure information for AnsibleTower

### CODE REVIEW

**Comment by JacobCallahan** [comment]
@rplevka can you check this out locally and see if it meets your needs? I can make adjustments as needed.

Also, if anyone else would like other important information emitted, please let me know.

File: README.md [hunk 1]
```diff
@@ -135,6 +135,14 @@ broker execute -o raw --workflow my-awesome-workflow --additional-arg True
 broker execute -o raw --workflow my-awesome-workflow --additional-arg True --artifacts last
 ```
 
+**machine processable output**
+
+If running in a CI or other automated environment, Broker offers the choice to store important output information in an output file. This is json-formatted data. Please be aware that any existing file with the matching path and name will be erased.
```
Comment (tstrych) [inline README.md:140]: can you be more specific what is important output information ?
when I read this I don't know what values I can expect inside the json file.

File: tests/test_helpers.py [hunk 2]
```diff
@@ -35,3 +36,16 @@ def test_resolve_file_args():
...
+    out_file = tmp_path / "output.json"
+    assert not out_file.exists()
+    helpers.emit.set_file(out_file)
+    assert out_file.exists()
+    helpers.emit(test="value", another=5)
+    written = json.loads(out_file.read_text())
+    assert written == {"test": "value", "another": 5}
+    helpers.emit({"thing": 13})
+    written = json.loads(out_file.read_text())
+    assert written == {"test": "value", "another": 5, "thing": 13}
```
Comment (tstrych) [inline tests/test_helpers.py:51]: I really appreciate that you wrote a test on these. Maybe I would also mock a bit more real data, just to see it on something more real and more complex. 
Of course not required change for emit v1 :+1:

File: broker/helpers.py [hunk 3]
```diff
@@ -249,6 +249,46 @@ def yaml_format(in_struct):
...
+            self.file.unlink()
+        self.file.touch()
+
+    def emit_to_file(self, *args, **kwargs):
+        if not self.file:
+            return
+        for arg in args:
+            if not isinstance(arg, dict):
+                raise exceptions.BrokerError(f"Received an invalid data emission {arg}")
+            kwargs.update(arg)
```
Comment (tstrych) [inline broker/helpers.py:280]: I don't understand the purpose of this line: `kwargs.update(arg)`
Would you mind explain it to me ?

**Review by tstrych — APPROVED** [review]
LGTM. Approving in front, but should be further discussed with @rplevka as Jake mentioned.

File: README.md (hunk 1 above)
Comment (JacobCallahan) [reply README.md:140]: I could flesh out the description a little if this PR needs other updates

File: tests/test_helpers.py (hunk 2 above)
Comment (JacobCallahan) [reply tests/test_helpers.py:51]: More complex tests are always welcome in future PRs :)

File: broker/helpers.py (hunk 3 above)
Comment (JacobCallahan) [reply broker/helpers.py:280]: Sure thing! So the emitter accepts both keyword-based arguments as well as positional arguments.
In this loop, we're going over a list of possible positional arguments, which can only be dictionaries.
We then update the keyword arguments dictionary with each of the positional dictionaries passed in.
This allows us to just track one object that will be used to update the existing data.

**Comment by rplevka** [comment]
We tried this with @ogajduse 
and any commands returns the same error instantly:
```
$ broker --help
[ERROR 210610 17:24:46] BrokerError: expected str, bytes or os.PathLike object, not NoneType
```

**Comment by rplevka** [comment]
so I tried to run a simple workflow - `list-templates` and broker crashed as `WorkflowJob`  seemed not being json-serializable:

```
Traceback (most recent call last):
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/broker/commands.py", line 18, in __call__
    return self.main(*args, **kwargs)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/broker/commands.py", line 372, in execute
    helpers.emit({"output": result})
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/broker/helpers.py", line 287, in __call__
    return self.emit_to_file(*args, **kwargs)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/broker/helpers.py", line 284, in emit_to_file
    self.file.write_text(json.dumps(curr_data, indent=4, sort_keys=True))
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/awxkit/api/mixins/has_create.py", line 186, in _filter_ds_from_payload
    return dumps(filtered, *a, **kw)
  File "/usr/lib64/python3.9/json/__init__.py", line 234, in dumps
    return cls(
  File "/usr/lib64/python3.9/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/lib64/python3.9/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/lib64/python3.9/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/lib64/python3.9/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/lib64/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type WorkflowJob is not JSON serializable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rplevka/.virtualenvs/broker_jake/bin/broker", line 8, in <module>
    sys.exit(cli())
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/broker/commands.py", line 22, in __call__
    helpers.emit(return_code=err.error_code, error_message=err.message)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/broker/helpers.py", line 287, in __call__
    return self.emit_to_file(*args, **kwargs)
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/broker/helpers.py", line 284, in emit_to_file
    self.file.write_text(json.dumps(curr_data, indent=4, sort_keys=True))
  File "/home/rplevka/.virtualenvs/broker_jake/lib/python3.9/site-packages/awxkit/api/mixins/has_create.py", line 186, in _filter_ds_from_payload
    return dumps(filtered, *a, **kw)
  File "/usr/lib64/python3.9/json/__init__.py", line 234, in dumps
    return cls(
  File "/usr/lib64/python3.9/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/lib64/python3.9/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/lib64/python3.9/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/lib64/python3.9/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/lib64/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type TypeError is not JSON serializable

```

File: broker/providers/ansible_tower.py [hunk 4]
```diff
@@ -374,20 +415,23 @@ def execute(self, **kwargs):
...
+        helpers.emit(job_url=job_ui_url)
         logger.info(
             "Waiting for job: \n"
             f"API: {url_parser.urljoin(self.url, str(job.url))}\n"
             f"UI: {job_ui_url}"
         )
         job.wait_until_completed(timeout=settings.ANSIBLETOWER.workflow_timeout)
         if not job.status == "successful":
+            message_data = {
+                f"{subject.capitalize()} Status:": job.status,
```
Comment (ogajduse) [inline broker/providers/ansible_tower.py:427]: Can we get rid of the colon here? It does not look nice in the JSON output. But this is just more a cosmetic thing.
```json
{
   "Workflow Status:": "failed",
}
```

**Review by ogajduse — COMMENTED** [review]
Good job @JacobCallahan!
I have a few comments to add.

I am completely satisfied with the JSON produced in case of success. We can not wish to display/output more than it provides. 
To the output produced in case of failure that is listed below: The job URLs listed in there are pointing to AT WebUI. I would like the JSON to also contain links to AT API so we would have a direct API link for the subsequent machine processing without the need for any API URL compilation.

```json
{
  "Reason(s)": {
    "job": "satlab-tower-deploy-base-rhel-from-rhvm-02",
    "reason": "No result was found for the passed string, or the string passed not found",
    "task": "Deploying a RHEL Template on RHVM"
  },
  "URL": "https://our.at.com/#/workflows/638255",
  "Workflow Status:": "failed",
  "action": "execute",
  "arguments": {
    "rhel_version": "7",
    "workflow": "deploy-base-rhel"
  },
  "auth_type": "token",
  "hosts": [],
  "job_url": "https://our.at.com/#/workflows/638255",
  "provider": "AnsibleTower"
}
```

I assume the implementation would like similar to this:
https://github.com/SatelliteQE/broker/blob/f6a5fc4c989f60e3c3f4af1dacfc132e65790b88/broker/providers/ansible_tower.py#L377-L381

The JSON output of the failed runs does not include `output`. Is it possible to include it? I see it as an extra Tower API call to `/api/v2/workflow_jobs/<jobid>/`. This will not add too much time, will it?

**Review by rplevka — CHANGES_REQUESTED** [review]
@JacobCallahan thanks for the implementation!
I definitely like the structure of the JSON, and I agree with @ogajduse that it could contain API urls in case of a failure.


ACK pending @ogajduse comments.

File: broker/hosts.py [hunk 5]
```diff
@@ -20,6 +20,19 @@ def __init__(self, hostname, name=None, from_dict=False, **kwargs):
         self.password = kwargs.get("pwassword", settings.HOST_PASSWORD)
         self.session = None
 
+    def __getstate__(self):
```
Comment (mshriver) [inline broker/hosts.py:23]: Nice!

**Review by mshriver — APPROVED** [review]

**Pull request merged**